synapse/storage/databases/main/event_push_actions.py

```diff
@@ -463,6 +465,153 @@ def add_thread_id_summary_txn(txn: LoggingTransaction) -> int:
 
         return result
 
+    async def get_unread_counts_by_room_for_user(self, user_id: str) -> Dict[str, int]:
+        """Get the notification count by room for a user. Only considers notifications,
+        no highlights or unreads, and threads are currently aggregated under their room.
+
+        This function is intentionally not cached because it is called to calculate the
+        unread badge for notifications and thus the result is expected to change.
+
+        Note that this function assumes the user is a member of the room. Because
+        summary rows are not removed when a user leaves a room, the caller must
+        filter out those results from the result.
```

Comment:
Is this a bug that Synapse doesn't delete those?

(Not saying we should fix it in this PR, just curious of your thoughts here!)

Reply:
Yeah I think it is! Added https://github.com/matrix-org/synapse/issues/14567

Reply:
Thanks! Let's leave that to the separate issue then.